cin_validator/rules/cin2022_23/rule_4016.py

```diff
@@ -0,0 +1,306 @@
+
+import pandas as pd
+
+from cin_validator.rule_engine import CINTable, RuleContext, rule_definition
+from cin_validator.test_engine import run_rule
+from cin_validator.utils import make_census_period
+
+# Get tables and columns of interest from the CINTable object defined in rule_engine/__api.py
+
+ChildProtectionPlans = CINTable.ChildProtectionPlans
+LAchildID = ChildProtectionPlans.LAchildID
+CPPID = ChildProtectionPlans.CPPID
+CPPstartDate = ChildProtectionPlans.CPPstartDate
+CPPendDate = ChildProtectionPlans.CPPendDate
+
+CINplanDates = CINTable.CINplanDates
+CINPlanStartDate = CINplanDates.CINPlanStartDate
+
+Header = CINTable.Header
+ReferenceDate = Header.ReferenceDate
+
+
+# define characteristics of rule
+@rule_definition(
+    # write the rule code here
+    code=4016,
+    # replace ChildProtectionPlans with the value in the module column of the excel sheet corresponding to this rule .
+    # Note that even if multiple tables are involved, one table will be named in the module column.
+    module=CINTable.CINplanDates,
+    # replace the message with the corresponding value for this rule, gotten from the excel sheet.
+    message="A CIN Plan has been reported as open at the same time as a Child Protection Plan.",
+    # The column names tend to be the words within the < > signs in the github issue description.
+    affected_fields=[
+        CINPlanStartDate,
+        CPPstartDate,
+        CPPendDate,
+    ],
+)
+def validate(
+    data_container: Mapping[CINTable, pd.DataFrame], rule_context: RuleContext
+):
+    # PREPARING DATA
+
+    # Replace ChildProtectionPlans with the name of the table you need.
+    df_cpp = data_container[ChildProtectionPlans].copy()
+    df_cin = data_container[CINplanDates].copy()
+
+    # Before you begin, rename the index so that the initial row positions can be kept intact.
+    df_cpp.index.name = "ROW_ID"
+    df_cin.index.name = "ROW_ID"
+
+    # Resetting the index causes the ROW_IDs to become columns of their respective DataFrames
+    # so that they can come along when the merge is done.
+    df_cpp.reset_index(inplace=True)
+    df_cin.reset_index(inplace=True)
+
+    # ReferenceDate exists in the header table so we get header table too.
+    df_ref = data_container[Header]
+    ref_date_series = df_ref[ReferenceDate]
+
+    # the make_census_period function generates the start and end date so that you don't have to do it each time.
+    collection_start, reference_date = make_census_period(ref_date_series)
+
+    # lOGIC
+    # Implement rule logic as described by the Github issue.
+    # Put the description as a comment above the implementation as shown.
+
+    # The <CINPlanStartDate> (N00689) for any CIN Plan group cannot fall within either:
+    # <CPPstartDate> (N00105) or <CPPendDate> (N00115);
+    # or <CPPstartDate> and <ReferenceDate> (N00603) if <CPPendDate> is not present - for any CPP group;
+    # unless <CINPlanStartDate> is equal to <CPPendDate> for this group
+
```

Comment:
Nope, on second thought don't do that here because the condition has an exception only for CPPendDate and not reference_date. So we'll need to know which one we're comparing CINplanStartDate with.

I'm referring to this clause: "unless CINPlanStartDate is equal to CPPendDate for this group"